File: examples2d/pin_slot_joint2.rs
use crate::utils::character;
use crate::utils::character::CharacterControlMode;
use rapier_testbed2d::TestbedViewer;
use rapier2d::control::{KinematicCharacterController, PidController};
use rapier2d::prelude::*;

pub async fn run(viewer: &mut TestbedViewer) -> anyhow::Result<()> {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Ground
     */
    let ground_size = 3.0;
    let ground_height = 0.1;

    let rigid_body_floor = RigidBodyBuilder::fixed().translation(Vector::new(0.0, -ground_height));
    let floor_collider = ColliderBuilder::cuboid(ground_size, ground_height);
    let _ = world.insert(rigid_body_floor, floor_collider);

    /*
     * Character we will control manually.
     */
    let character_pos = Vector::new(0.0, 0.3);
    let rigid_body_character =
        RigidBodyBuilder::kinematic_position_based().translation(character_pos);
    let character_collider = ColliderBuilder::cuboid(0.15, 0.3);
    let (character_handle, _) = world.insert(rigid_body_character, character_collider);

    /*
     * Tethered cube.
     */
    let rad = 0.4;
    let slot_axis = Vector::new(1.0, 1.0).normalize();
    let slot_anchor = Vector::new(2.0, 2.0);
    let slot_limits = [-1.0, f32::INFINITY];
    // Start the cube on the slot's lower limit so the joint doesn't have to push it back.
    let cube_pos = character_pos + slot_anchor + slot_axis * slot_limits[0] - Vector::new(0.0, rad);

    let rigid_body_cube = RigidBodyBuilder::new(RigidBodyType::Dynamic).translation(cube_pos);
    let cube_collider = ColliderBuilder::cuboid(rad, rad);
    let (cube_handle, _) = world.insert(rigid_body_cube, cube_collider);

    /*
     * SimdRotation axis indicator ball.
     */
    let rigid_body_ball =
        RigidBodyBuilder::new(RigidBodyType::Dynamic).translation(cube_pos + Vector::new(0.0, rad));
    let ball_collider = ColliderBuilder::ball(0.1);
    let (ball_handle, _) = world.insert(rigid_body_ball, ball_collider);

    /*
     * Fixed joint between rotation axis indicator and cube.
     */
    let fixed_joint = FixedJointBuilder::new()
        .local_anchor1(Vector::new(0.0, 0.0))
        .local_anchor2(Vector::new(0.0, -rad))
        .build();
    world.insert_impulse_joint(cube_handle, ball_handle, fixed_joint);

    /*
     * Pin slot joint between cube and ground.
     */
    let pin_slot_joint = PinSlotJointBuilder::new(slot_axis)
        .local_anchor1(slot_anchor)
        .local_anchor2(Vector::new(0.0, rad))
        .limits(slot_limits) // Set the limits for the pin slot joint
        .build();
    world.insert_impulse_joint(character_handle, cube_handle, pin_slot_joint);

    /*
     * Callback to update the character based on user inputs.
     */
    let mut control_mode = CharacterControlMode::Kinematic(0.1);
    let mut controller = KinematicCharacterController::default();
    let mut pid = PidController::default();

    /*
     * Set up the testbed.
     */
    viewer.set_world(&mut world);
    viewer.look_at(Vec2::new(0.0, 1.0), 100.0);

    while viewer.render_frame(&mut world).await {
        if viewer.simulating() {
            world.step();
            character::update_character(
                viewer,
                &mut world,
                &mut control_mode,
                &mut controller,
                &mut pid,
                character_handle,
            );
        }
    }
    Ok(())
}

File: python/rapier-testbed/rapier_testbed/examples2/pin_slot_joint2.py
"""Port of examples2d/pin_slot_joint2.rs.

The Rust version drives the character with the keyboard via
``utils::character``. The Python port keeps the static scene (pin-slot
joint + tethered cube) but omits the keyboard-driven controller.
"""
from __future__ import annotations

import math

import rapier2d as rp
from .._registry import register

CATEGORY = "Joints"
NAME = "Pin Slot Joint"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 3.0
    ground_height = 0.1

    floor_handle = bodies.insert(
        rp.RigidBody.fixed().translation((0.0, -ground_height))
    )
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height), floor_handle, bodies
    )

    character_handle = bodies.insert(
        rp.RigidBody.kinematic_position_based().translation((0.0, 0.3))
    )
    colliders.insert_with_parent(
        rp.Collider.cuboid(0.15, 0.3), character_handle, bodies
    )

    rad = 0.4
    cube_handle = bodies.insert(rp.RigidBody.dynamic().translation((1.0, 1.0)))
    colliders.insert_with_parent(rp.Collider.cuboid(rad, rad), cube_handle, bodies)

    ball_handle = bodies.insert(rp.RigidBody.dynamic().translation((1.0, 1.0)))
    colliders.insert_with_parent(rp.Collider.ball(0.1), ball_handle, bodies)

    fixed_joint = (
        rp.FixedJoint.builder()
        .local_anchor1((0.0, 0.0))
        .local_anchor2((0.0, -0.4))
    )
    impulse_joints.insert(cube_handle, ball_handle, fixed_joint, wake_up=True)

    axis_len = math.sqrt(2.0)
    axis = (1.0 / axis_len, 1.0 / axis_len)
    pin_slot_joint = (
        rp.PinSlotJoint.builder(axis)
        .local_anchor1((2.0, 2.0))
        .local_anchor2((0.0, 0.4))
        .limits(-1.0, float("inf"))
    )
    impulse_joints.insert(
        character_handle, cube_handle, pin_slot_joint, wake_up=True
    )

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.set_camera_2d(center=(0.0, 1.0), zoom=100.0)


register(CATEGORY, NAME, init_world, dim=2)

if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
